Core Layout Support + Many, many additions.

avaidyam opened the pull request:
Adding Core Layout (Cocoa Auto Layout) support to TwUI. Similar, but more inspired by CAConstraint at the moment. In later commits, I'll be adding more NSLayoutConstraint features to it. I've also added some TUIControl additions to mimic iOS's ease of building controls (the background tracking is taken care of, stick to the methods begin/continue/end/cancel tracking, and all will be well). TUIButton has two or three small things, and it now takes 6 lines to get a fully functioning button positioned and everything. TUIColor additions are just nifty tricks, not sure where I picked them up though.

I also renamed "touch" to "mouse" in the TUIControl events, and added a few more.

Some issues with constraints:
- It's not based on CAConstraints. Rewritten from scratch to think like a hybrid between that and NSLayoutConstraint.
- Does NOT like animations. Not sure how to get this to work, perhaps look into CAMediaTiming and adjust periodically?
- Many features don't work yet, like priorities, inequalities, and two or three attributes.
- I just found out about an ancient project called Cassowary that OS X Lion actually uses to support Auto Layout! It's open source too! I'll have to look into this, but I'll probably rewrite this as a wrapper to CAConstraints to avoid the hassles, and stick to Core Animation.

avaidyam committed:
Core Layout Support

Adding Core Layout (Cocoa Auto Layout) support to TwUI. Similar, but
more inspired by CAConstraint at the moment. In later commits, I'll be
adding more NSLayoutConstraint features to it.

joshaber commented:
It looks like this can't be merged right now without conflicts. Could you resolve those please?

avaidyam commented:
Oh wow okay, I tested this on yesterday's github/dev branch, but I'll patch it up :]

joshaber commented:
:) moving fast!

avaidyam commented:
Yep! That's a good sign :D

avaidyam committed:
Revert "Core Layout Support"

This reverts commit a5013e7e13d3927e636d0332894f29955108a5ae.

avaidyam committed:
Revert "Revert "Core Layout Support""

This reverts commit 9c65ce3ded9d4596e0f4f1e97f0edb8aed048fba.

avaidyam commented:
I can't figure out for the life of me where the conflicts arise... :\ It seems fine to me, but then again, I'm also a complete newb to git/github.

avaidyam committed:
Added more features

Button and Color features- picked up some nifty Color modification
methods but not sure from where.

avaidyam commented:
EDIT: Fixed the issues, seems like Xcode decided not to tell me about ARC incompatibility.  Fixed those issues, added:

TUIButton stuff (create a button that works perfectly in 6 lines now!).
TUIColor modification methods (not sure where I found them, but I use them all extensively, so it seemed fine to add them...).
TUIControl tracking methods. Fiddle not with mouseDown and mouseUp methods! Use the familiar begin/continue/end/cancel tracking methods with controls!

So yeah, figured out where it went wrong. I think it should merge properly now. I've got millions more additions too! ;D

avaidyam committed:
Bug fixes galore!

Should merge properly now. Contains new TUIButton,  TUIColor, and
TUIControl methods.

joshaber commented:
Looks like it still can't merge cleanly :(

avaidyam commented:
Can you tell me where the issues are so I can fix them? Sorry!

joshaber commented:
Hmm, I thought this was the autolayout stuff but it looks like there are a bunch more changes in the pull request now?

avaidyam commented:
I'm sorry D: I'm new to github and I tried adding more of my private fork stuff back in and it merged into this pull request. But the things I did add are new TUIControl features (begin/continue/end/cancel tracking with event methods) and some more TUIColors and color modifications. Dunno what to do now...

avaidyam commented:
How do I do an upstream merge? Maybe that'll fix the issues?

avaidyam committed:
Merge remote-tracking branch 'upstream/dev' into dev

Conflicts:
	lib/UIKit/TUIView+Private.h
	lib/UIKit/TUIView.m

avaidyam commented:
EDIT: Finally fixed the merge issues! Really trivial ones too.... hopefully it's okay to pull now?

avaidyam committed:
Whoops- added a 'static'.

No. more. bugs. .____.

joshaber commented on lib/UIKit/TUIView.m:
```diff
@@ -1066,3 +1095,296 @@ - (BOOL)eventInside:(NSEvent *)event
+    static BOOL configured = NO;
+    if(configured) return;
+    
+    if(![NSStringFromClass(self.class) hasPrefix:@"TUILayoutConstraint_"]) {
```
Is there a reason we need to do this dynamic subclassing? Where does `-initializeInternal` get called from?

avaidyam replied:
Oh, haha, that should be commented out right here, so it won't be called. In case you wanted to use this with a UIView, you'd use associated objects and insert [self initializeInternal]; to set it up. For TwUI, we have the source, so it's already been taken care of.

joshaber replied:
Ah cool, can we just remove them then?

joshaber commented on lib/UIKit/TUIView.m:
```diff
@@ -57,7 +65,25 @@ - (TUIView *)closestAssociatedView
+
 @property (nonatomic, strong) NSMutableArray *subviews;
+
+@property (nonatomic, retain) NSMutableArray *liveConstraints;
```
strong instead of retain, please

avaidyam committed:
ARC issues

avaidyam commented:
Well, it should be good to go now...

avaidyam commented:
Well, since this hasn't been pulled yet, I'm going to close this, and create a new fork/pull request for the newer model of Constraints I've built. Now allows NSValueTransformers and block-based constraints! Still need to fix up the animation issues though. Also: new TUIPanel, and TUIPopover makeover, which looks JUST like the Lion NSPopover, but no custom background view. Instead, I'm adding customization drawing blocks/colors/etc.

joshaber commented:
Cool, I'll look forward to that PR!

avaidyam commented:
It's in!
